src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphDAO.scala

```diff
@@ -135,53 +169,125 @@ trait GraphDAO {
+   * special function passed into idFxn of serializeDomainObjects when there should only be
+   * one object and thus any existing should be overwritten
    */
-  def fromVertex[T: TypeTag](vertex: Vertex, otherProperties: Map[String, Any]): T = {
-    fromPropertyMap(otherProperties ++ getVertexProperties(vertex))
+  private def thereShouldOnlyBeOne(x: Any) = 1
+
+  /**
+   * Serializes rawls domain objects to sub vertices. Will overwrite corresponding existing vertices, correspondence
+   * determined by idFxn.
+   *
+   * @param vertex the top level vertex
+   * @param objs the nested objects
+   * @param edgeLabel the label to give to the edges to the sub vertices
+   * @param idFxn function mapping an object of type T to an identifier, used to determine if vertices already in the
+   *              graph correspond to any of the nested objects
+   * @param graph used to create new vertices
+   * @param workspaceName used to construct domain objects that require a WorkspaceName
+   * @tparam T the type of the object being serializes
+   */
+  private def serializeDomainObjects[T: TypeTag: ClassTag](vertex: Vertex, objs: Seq[T], edgeLabel: String, idFxn: T => Any, graph: Graph, workspaceName: WorkspaceName): Seq[Vertex] = {
+    val existingObjVertexesById = vertex.getVertices(Direction.OUT, edgeLabel).map { objVertex =>
+      val obj = loadFromVertex[T](objVertex, Option(workspaceName))
+      idFxn(obj) -> objVertex
+    } toMap
+
+    (existingObjVertexesById -- objs.map(idFxn)).foreach(_._2.remove())
+
+    objs.map { obj =>
+      val objVertex = existingObjVertexesById.getOrElse(idFxn(obj), {
+        val newVertex = addVertex(graph, VertexSchema.vertexClassOf[T])
+        vertex.addEdge(edgeLabel, newVertex)
+        newVertex
+      })
+
+      saveToVertex(obj, objVertex, graph, workspaceName)
+    }
+  }
+
+  private def serializeAttributeMap(vertex: Vertex, propName: String, map: Map[String, Attribute], graph: Graph, workspaceName: WorkspaceName): Unit = {
+    // remove existing map then repopulate
+    vertex.getVertices(Direction.OUT, propName).headOption.foreach(removeMapVertex)
+
+    val mapVertex = addVertex(graph, VertexSchema.Map)
+    vertex.addEdge(propName, mapVertex)
+
+    map.foreach { entry =>
+      entry._2 match {
+        case v: AttributeValue => mapVertex.setProperty(entry._1, AttributeConversions.attributeToProperty(v))
+        case ref: AttributeReference => serializeReference(mapVertex, entry._1, ref, graph, workspaceName)
+        case AttributeValueList(values) => serializeAttributeMap(mapVertex, entry._1, values.zipWithIndex.map(e => e._2.toString -> e._1).toMap, graph, workspaceName)
+        case AttributeReferenceList(references) => serializeAttributeMap(mapVertex, entry._1, references.zipWithIndex.map(e => e._2.toString -> e._1).toMap, graph, workspaceName)
+        case AttributeEmptyList => serializeAttributeMap(mapVertex, entry._1, Map.empty, graph, workspaceName)
+      }
```

Comment:
The ```case (key, value)``` syntax would be nicer than ```entry._1, entry._2```, especially since we have other tuple operations nested inside here.